exp/state/remote_state.go

```diff
@@ -0,0 +1,223 @@
+
+	if componentType == "accounts" {
+		if len(accounts) > 0 {
+			c := conf.Accounts[componentName]
+			if c.Common.DependsOn == nil {
+				c.Common.DependsOn = &v2.DependsOn{}
+			}
+
+			c.DependsOn.Accounts = accounts
+			conf.Accounts[componentName] = c
+		}
+	} else if componentType == "env" {
+		if len(accounts) > 0 || len(components) > 0 {
+			c := conf.Envs[envName].Components[componentName]
+
+			if c.Common.DependsOn == nil {
+				c.Common.DependsOn = &v2.DependsOn{}
+			}
+
+			c.DependsOn.Accounts = accounts
+			c.DependsOn.Components = components
+
+			conf.Envs[envName].Components[componentName] = c
+		}
+	}
+
+	return conf.Write(fs, configFile)
+}
+
+func collectRemoteStateReferences(path string) ([]string, error) {
+	fs := tfconfig.NewOsFs()
+
+	references := map[string]bool{}
+
+	primaryPaths, err := dirFiles(fs, path)
+	if err != nil {
+		return nil, err
+	}
+
+	parser := hclparse.NewParser()
+
+	for _, filename := range primaryPaths {
+		logrus.Debugf("reading file %s", filename)
+		b, err := fs.ReadFile(filename)
+		if err != nil {
+			return nil, err
+		}
+
+		var file *hcl.File
+		var fileDiags hcl.Diagnostics
+
+		if strings.HasSuffix(filename, ".json") {
+			file, fileDiags = parser.ParseJSON(b, filename)
+		} else {
+			file, fileDiags = parser.ParseHCL(b, filename)
+		}
+		if fileDiags.HasErrors() {
+			return nil, fileDiags
+		}
+
+		if file == nil {
+			continue
+		}
+
+		content, _, contentDiags := file.Body.PartialContent(rootSchema)
+		if contentDiags.HasErrors() {
+			return nil, contentDiags
+		}
+
+		logrus.Debugf("len(content.Blocks) %v", len(content.Blocks))
+		for _, block := range content.Blocks {
+			logrus.Debugf("block type: %v", block.Type)
+
+			attrs, _ := block.Body.JustAttributes()
+
+			for _, v := range attrs {
+				refs, _ := lang.ReferencesInExpr(v.Expr)
+
+				for _, r := range refs {
+					if r != nil {
+						logrus.Debugf("ref: %v", r)
+						if resource, ok := r.Subject.(addrs.ResourceInstance); ok {
+							if resource.Resource.Type == "terraform_remote_state" {
+								references[resource.Resource.Name] = true
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	refNames := []string{}
+	for k := range references {
+		refNames = append(refNames, k)
+	}
+
+	sort.Strings(refNames)
+	return refNames, nil
+}
+
+// taken from https://github.com/hashicorp/terraform-config-inspect/blob/c481b8bfa41ea9dca417c2a8a98fd21bd0399e14/tfconfig/load.go#L81
+func dirFiles(fs tfconfig.FS, dir string) ([]string, error) {
+	var primary []string
+
+	infos, err := fs.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var override []string
+	for _, info := range infos {
+		if info.IsDir() {
+			// We only care about files
+			continue
+		}
+
+		name := info.Name()
+		ext := fileExt(name)
+		if ext == "" || isIgnoredFile(name) {
+			continue
+		}
+
+		baseName := name[:len(name)-len(ext)] // strip extension
+		isOverride := baseName == "override" || strings.HasSuffix(baseName, "_override")
```

Comment:
huh, I didn't know about this behavior in TF

Reply:
yeah it is weird